openmc/data/resonance.py

```diff
@@ -0,0 +1,1063 @@
+    energy_min : float
+        Minimum energy of the resolved resonance range in eV
+    energy_max : float
+        Maximum energy of the resolved resonance range in eV
+    channel : dict
+        Dictionary whose keys are l-values and values are channel radii as a
+        function of energy
+    scattering : dict
+        Dictionary whose keys are l-values and values are scattering radii as a
+        function of energy
+
+    Attributes
+    ----------
+    channel_radius : dict
+        Dictionary whose keys are l-values and values are channel radii as a
+        function of energy
+    energy_max : float
+        Maximum energy of the resolved resonance range in eV
+    energy_min : float
+        Minimum energy of the resolved resonance range in eV
+    scattering_radius : dict
+        Dictionary whose keys are l-values and values are scattering radii as a
+        function of energ
+    target_spin : float
+        Intrinsic spin, :math:`I`, of the target nuclide
+
+    """
+    def __init__(self, target_spin, energy_min, energy_max, channel, scattering):
+        self.target_spin = target_spin
+        self.energy_min = energy_min
+        self.energy_max = energy_max
+        self.channel_radius = channel
+        self.scattering_radius = scattering
+
+        self._prepared = False
+        self._parameter_matrix = {}
+
+    @classmethod
+    def from_endf(cls, ev, file_obj, items):
+        """Create resonance range from an ENDF evaluation.
+
+        This factory method is only used when LRU=0, indicating that only a
+        scattering radius appears in MF=2, MT=151. All subclasses of
+        ResonanceRange override this method with their own.
+
+        Parameters
+        ----------
+        ev : openmc.data.endf.Evaluation
+            ENDF evaluation
+        file_obj : file-like object
+            ENDF file positioned at the second record of a resonance range
+            subsection in MF=2, MT=151
+        items : list
+            Items from the CONT record at the start of the resonance range
+            subsection
+
+        Returns
+        -------
+        openmc.data.ResonanceRange
+            Resonance range data
+
+        """
+        energy_min, energy_max = items[0:2]
+
+        # For scattering radius-only, NRO must be zero
+        assert items[4] == 0
+
+        # Get energy-independent scattering radius
+        items = get_cont_record(file_obj)
+        target_spin = items[0]
+        ap = Polynomial((items[1],))
+
+        # Calculate channel radius from ENDF-102 equation D.14
+        a = Polynomial((0.123 * (1.008665*ev.target['mass'])**(1./3.) + 0.08,))
+
+        return cls(target_spin, energy_min, energy_max, {0: a}, {0: ap})
+
+    def reconstruct(self, energies):
+        """Evaluate cross section at specified energies.
+
+        Parameters
+        ----------
+        energies : float or Iterable of float
+            Energies at which the cross section should be evaluated
```

Comment:
There is still quite a ways to go before we could actually use this to generate HDF5 data for the transport solver, so I'm not too concerned about that for now. I'm kind of hoping to defer on the units issue because I didn't want this PR to grow too huge. I can envision a future PR just dealing with units issues. As far as the use case, in the short term I would like to use this to generate 0K elastic scattering for DBRC.

As far as MeV versus eV, I find eV to be the more "natural" unit. To me, ACE/MCNP are kind of an oddity being in MeV. I'm sure there are reasons they chose to use MeV but I'd take a guess that it has to do with their application area. And, ultimately, if we really want to go from ENDF/GND -> HDF5, it doesn't make sense to convert to MeV. Of course, if we do decide to eventually adopt eV units, users will need a fair bit of warning.

Regarding the second option (have data types know about their units), I completely agree that on the Fortran side we would *not* want to support that, but it wouldn't be as terrible on the Python side. I bring this up because the data abstractions in GND do have units listed explicitly. It's conceivable then that a future evaluation might contain data using non-standard units because the format allows it. You might say "no, evaluators wouldn't do something silly like that", but I've learned over time that evaluators do plenty of silly things.